openedx/core/djangoapps/programs/tasks.py

```diff
@@ -782,39 +783,52 @@ def update_certificate_available_date_on_course_update(self, course_key):
-        )
-        update_credentials_course_certificate_configuration_available_date.delay(
-            str(course_key), str(course_overview.certificate_available_date)
-        )
-    # OR,
-    #   - The course-run is self-paced, AND
-    #   - The `certificate_available_date` is (now) None. (This task will be executed after an update to the course
-    #     overview)
-    # There are times when the CourseCertificate configuration of a self-paced course-run in Credentials can become
-    # associated with a `certificate_available_date`. This ends up causing learners' certificate to be incorrectly
-    # hidden. This is due to the Credentials IDA not understanding the concept of course pacing. Thus, we need a way
-    # to remove this value from self-paced courses in Credentials.
-    elif course_overview and course_overview.self_paced is True and course_overview.certificate_available_date is None:
-        LOGGER.info(
-            "Queueing task to remove the `certificate_available_date` in the Credentials service for course-run "
-            f"{course_key}"
-        )
+    # When updating the certificate available date of instructor-paced course runs,
+    #   - If the display behavior is set to "A date after the course end date" (END_WITH_DATE), we should send the
+    #     certificate available date set by the course team in Studio (and stored as part of the course runs Course
+    #     Overview)
+    #   - If the display behavior is set to "End date of course" (END), we should send the end date of the course run
+    #     as the certificate available date. We send the end date because the Credentials IDA doesn't understand the
+    #     concept of course pacing and needs an explicit date in order to correctly gate the visibility of course and
+    #     program certificates.
+    #   - If the display behavior is set to "Immediately upon passing" (EARLY_NO_INFO), we should always send None for
+    #     the course runs certificate available date. A course run configured with this display behavior must not have a
+    #     certificate available date associated with or the Credentials system will incorrectly hide certificates from
+    #     learners.
+    course_overview = get_course_overview_or_none(course_key)
+    if (course_overview and course_overview.self_paced is False):
+        if course_overview.certificates_display_behavior == CertificatesDisplayBehaviors.END_WITH_DATE:
+            update_credentials_course_certificate_configuration_available_date.delay(
+                str(course_key),
+                str(course_overview.certificate_available_date)
+            )
+        elif course_overview.certificates_display_behavior == CertificatesDisplayBehaviors.END:
+            update_credentials_course_certificate_configuration_available_date.delay(
+                str(course_key),
+                str(course_overview.end)  # `end_date` is deprecated, use `end` instead
+            )
```

Comment:
### New behavior

For awhile now, we've had issues that can be boiled down to "Credentials doesn't understand course pacing". An update to a course run using this type of course pacing would cause no update to be sent to Credentials. 

This means that if a course run went from a display behavior of "END WITH DATE", using a certificate available date, and later updated to a display behavior "END", we would not clean up a leftover certificate available date in Credentials.

After thinking about it for a bit, I realized that Credentials doesn't NEED to understand course pacing to handle this type of configuration. If a course run is configured with a display behavior of "END", then the certificate available date is _the end date of the course_. 

With this change, we will start syncing the end date of the course run as the certificate available date in Credentials for courses configured with a display behavior of "END".